File: infra/src/main/java/gg/essential/cosmetics/model/CosmeticAssets.java
/*
 * Copyright (c) 2024 ModCore Inc. All rights reserved.
 *
 * This code is part of ModCore Inc.'s Essential Mod repository and is protected
 * under copyright registration # TX0009138511. For the full license, see:
 * https://github.com/EssentialGG/Essential/blob/main/LICENSE
 *
 * You may not use, copy, reproduce, modify, sell, license, distribute,
 * commercialize, or otherwise exploit, or create derivative works based
 * upon, this file or any other in this repository, all of which is reserved by Essential.
 */
package gg.essential.cosmetics.model;

import gg.essential.lib.gson.annotations.SerializedName;
import gg.essential.model.EssentialAsset;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class CosmeticAssets {

    @SerializedName("a")
    private final @NotNull EssentialAsset thumbnail;

    @SerializedName("b")
    private final @Nullable EssentialAsset texture;

    @SerializedName("c")
    private final @NotNull CosmeticGeometry geometry;

    @SerializedName("d")
    private final @Nullable EssentialAsset animations;

    @SerializedName("e")
    private final @Nullable CosmeticSkinMask skinMask;

    @SerializedName("f")
    private final @Nullable EssentialAsset settings;

    public CosmeticAssets(
            final @NotNull EssentialAsset thumbnail,
            final @Nullable EssentialAsset texture,
            final @NotNull CosmeticGeometry geometry,
            final @Nullable EssentialAsset animations,
            final @Nullable CosmeticSkinMask skinMask,
            final @Nullable EssentialAsset settings
    ) {
        this.thumbnail = thumbnail;
        this.texture = texture;
        this.geometry = geometry;
        this.animations = animations;
        this.skinMask = skinMask;
        this.settings = settings;
    }

    public @NotNull EssentialAsset getThumbnail() {
        return this.thumbnail;
    }

    public @Nullable EssentialAsset getTexture() {
        return this.texture;
    }

    public @NotNull CosmeticGeometry getGeometry() {
        return this.geometry;
    }

    public @Nullable EssentialAsset getAnimations() {
        return this.animations;
    }

    public @Nullable CosmeticSkinMask getSkinMask() {
        return this.skinMask;
    }

    public @Nullable EssentialAsset getSettings() {
        return this.settings;
    }

}

File: subprojects/cosmetics/src/commonMain/kotlin/gg/essential/mod/cosmetics/CosmeticAssets.kt
/*
 * Copyright (c) 2024 ModCore Inc. All rights reserved.
 *
 * This code is part of ModCore Inc.'s Essential Mod repository and is protected
 * under copyright registration # TX0009138511. For the full license, see:
 * https://github.com/EssentialGG/Essential/blob/main/LICENSE
 *
 * You may not use, copy, reproduce, modify, sell, license, distribute,
 * commercialize, or otherwise exploit, or create derivative works based
 * upon, this file or any other in this repository, all of which is reserved by Essential.
 */
package gg.essential.mod.cosmetics

import gg.essential.mod.EssentialAsset
import gg.essential.model.Side
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient

@Serializable
data class CosmeticAssets(val allFiles: Map<String, EssentialAsset>) {
    @Transient
    private val files = object {
        val unknown = allFiles.toMutableMap()

        operator fun get(path: String): EssentialAsset? {
            unknown.remove(path)
            return allFiles[path]
        }
    }

    val thumbnail: EssentialAsset? = files["thumbnail.png"]
    val texture: EssentialAsset? = files["texture.png"]
    val emissiveTexture: EssentialAsset? = files["emissive.png"]
    val geometry: Geometry = Geometry(files["geometry.steve.json"] ?: EssentialAsset.EMPTY, files["geometry.alex.json"])
    val animations: EssentialAsset? = files["animations.json"]
    val defaultSkinMask: SkinMask = SkinMask(files["skin_mask.steve.png"], files["skin_mask.alex.png"])
    val sidedSkinMasks: Map<Side, SkinMask> = Side.values().associateWith { side ->
        val sideId = side.name.lowercase()
        SkinMask(files["skin_mask.steve.$sideId.png"], files["skin_mask.alex.$sideId.png"])
    }
    val settings: EssentialAsset? = files["settings.json"]
    val particles: Map<String, EssentialAsset> = allFiles.mapNotNull { (path, asset) ->
        if (path.startsWith("particles/") && path.endsWith(".json")) {
            files.unknown.remove(path)
            path to asset
        } else null
    }.toMap()
    val soundDefinitions: EssentialAsset? = files["sounds/sound_definitions.json"]

    val otherFiles: Map<String, EssentialAsset>
        get() = files.unknown

    @Serializable
    data class Geometry(
        val steve: EssentialAsset,
        val alex: EssentialAsset?,
    )

    @Serializable
    data class SkinMask(
        val steve: EssentialAsset?,
        val alex: EssentialAsset?,
    )
}